crates/wasi-provider/src/lib.rs

```diff
@@ -205,45 +224,49 @@ impl WasiRuntime {
         args: Vec<String>,
         dirs: HashMap<String, Option<String>>,
         log_file_location: L,
-    ) -> Result<Self, failure::Error> {
-        debug!("before parse file: {:?}", std::env::current_dir().unwrap());
+    ) -> Result<(Self, BufReader<File>), failure::Error> {
         let module_data = wat::parse_file(module_path)?;
-        debug!("after parse file");
 
         // We need to use named temp file because we need multiple file handles
         // and if we are running in the temp dir, we run the possibility of the
         // temp file getting cleaned out from underneath us while running. If we
         // think it necessary, we can make these permanent files with a cleanup
         // loop that runs elsewhere. These will get deleted when the reference
         // is dropped
-        let stdout = NamedTempFile::new_in(&log_file_location)?;
-        let stderr = NamedTempFile::new_in(&log_file_location)?;
-        debug!("after files");
-        Ok(WasiRuntime {
-            module_data,
-            env,
-            args,
-            dirs,
-            stdout,
-            stderr,
-        })
+        //
+        // As warned in the BufReader docs, creating multiple BufReaders on the
+        // same stream can cause data loss. So reopen a new file object so we
+        // don't drop any data
+        let output = NamedTempFile::new_in(&log_file_location)?;
+        let file_handle = output.reopen()?;
```

Comment:
When you create a new temp file, it does not appear to be wrapped in a buffered reader. And `output` is only used once here. So re-opening seems to be unnecessary. The BufReader docs just warn that you shouldn't have two buffered readers on the same file handle because they can independently advance the underlying reader.

Reply:
This is the file handle for the returned BufReader (which I think you'd like to change anyway). Otherwise I have already moved `output` when I put it in the struct. But let's see how it looks after I address the others